Enabled multi-gpu training, buffers, grad clip in vocoder, saving optimizer state, and more fixes

TheButlah opened the pull request:
+ Added `voc_clip_grad_norm` hparam
+ Enabled gradient clipping in vocoder
+ Added warning when gradient is nan in vocoder and synthesizer
+ Added workaround for broken nn.DataParallel in `utils.__init__.py`
* Refactored `tacotron.py` and `fatchord_version.py` to store step
  in a PyTorch buffer instead of a gradient-less parameter
* Refactored training code for WaveRNN and tacotron to automagically
  use data parallelism when in the presence of multiple GPUS.
! Note that your batch size must be divisible by the number of GPUS

TheButlah commented:
Let me know if you have any questions. I am less familiar with DataParallel but it seems to be working on my multi-gpu setup. Note that I did not enable multi-gpus for the `generate()` methods as those methods seemed non-trivial to add the code to. Also, as before, I have not messed with the notebooks at all.

TheButlah commented:
Also, let me know if the whole `data_parallel_workaround()` technique is no longer needed. It would be great if we could just use the normal `nn.DataParallel` instead.

TheButlah commented:
@fatchord let me know if there is anything you need to help get this merged!

fatchord commented:
@TheButlah Apologies for the delay! I hope to get around to testing this at the weekend. 

And thanks btw - this looks good.

fatchord commented:
Sorry one thing - would it possible to set the gpu ids in hparams.py or perhaps through argparse using that workaround?

TheButlah commented:
@fatchord I considered that, but I think the best way to do that is using the `CUDA_VISIBLE_DEVICES` flag. In general most people want to use all gpus available to train on, and having to explicitly specify which GPUs are used as a list seems like uneccessary configuration for most situations. Someone needing that level of control should know about `CUDA_VISIBLE_DEVICES` anyway.

TheButlah commented:
Update: I fixed a bug in my patch where I forgot to import one of the functions I made. More importantly though, I made the code save the optimizer state along with the latest weights in a separate file. This will ensure that when resuming training, the optimizer's momentum and other stateful properties are restored, ensuring that you can start and stop the model without affecting the properties of the training process. See https://github.com/fatchord/WaveRNN/issues/87#issuecomment-499456228

TheButlah commented:
OK This should be fixed and ready to go now. I identified a number of other bugs. Sorry for all the force pushes!

fatchord commented:
@TheButlah That all looks great thanks. I tried training the wavernn model on a multi-gpu server yesterday and that worked well. However when I tried tacotron it only used one gpu even though 3 more were available. Can you replicate that on your end? 

Re:setting devices - I think this could be useful, for example in my situation I am sometimes sharing gpus on the same server with my work colleagues so the flexibility would be great. I don't mind doing that part myself though.

fatchord commented:
I tried training tacotron from your fork and got this:

Traceback (most recent call last):
  File "train_tacotron.py", line 180, in <module>
    tts_train_loop(model, optimizer, train_set, lr, training_steps, attn_example)
  File "train_tacotron.py", line 39, in tts_train_loop
    m1_hat, m2_hat, attention = data_parallel_workaround(model, x, m)
  File "/code/ollie/test_multigpu/WaveRNN/utils/__init__.py", line 21, in data_parallel_workaround
    replicas = torch.nn.parallel.replicate(model, device_ids)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/replicate.py", line 112, in replicate
    buffer_copies_not_rg = _broadcast_coalesced_reshape(buffers_not_rg, devices, detach=True)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/replicate.py", line 76, in _broadcast_coalesced_reshape
    return comm.broadcast_coalesced(tensors, devices)
  File "/usr/local/lib/python3.6/dist-packages/torch/cuda/comm.py", line 39, in broadcast_coalesced
    return torch._C._broadcast_coalesced(tensors, devices, buffer_size)
RuntimeError: all tensors must be on devices[0]

fatchord commented:
Also I needed to add 
self.rnn1.flatten_parameters()
self.rnn2.flatten_parameters()
to fatchord_version.py to get wavernn to work (worked perfectly after that)

TheButlah commented:
I'll take a look today/this weekend! Also I tested wavernn myself and didn't seem to have any issues with it, what was the cause and effect of doing flatten_parameters?

fatchord commented:
@TheButlah For some reason it won't work without me using flatten_parameters() - not sure what it does tbh.

There's actually a couple of other bugs I wanna fix - will i do a pr on your fork and add the fixes there?

TheButlah commented:
I'll add you as a contributor, feel free to commit directly

On Fri, Jul 19, 2019, 10:30 fatchord <notifications@github.com> wrote:

> @TheButlah <https://github.com/TheButlah> For some reason it won't work
> without me using flatten_parameters() - not sure what it does tbh.
>
> There's actually a couple of other bugs I wanna fix - will i do a pr on
> your fork and add the fixes there?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fatchord/WaveRNN/pull/109?email_source=notifications&email_token=ABVFQR2K7LV43TRJ2IHWI4DQAH2ZVA5CNFSM4ID3ZNGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2MIFXQ#issuecomment-513311454>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABVFQRZBETLIT5CCHTYZZKDQAH2ZVANCNFSM4ID3ZNGA>
> .
>

TheButlah commented:
Just to confirm, are you using pytorch 1.1.0? Pytorch 1.0.0 is v different from 0.4 and that could be why

fatchord commented:
Yeah, I'm using torch 1.1.0

fatchord commented:
Apologies, I'm running 1.1 locally (single gpu) and nightly on the multi-gpu server - I'll see if I can get 1.1 running on the server.

TheButlah commented on gen_wavernn.py:
```diff
@@ -61,7 +61,7 @@ def gen_from_file(model, load_path, save_path, batched, target, overlap):
     parser.add_argument('--overlap', '-o', type=int, help='[int] number of crossover samples')
     parser.add_argument('--file', '-f', type=str, help='[string/path] for testing a wav outside dataset')
     parser.add_argument('--weights', '-w', type=str, help='[string/path] checkpoint file to load weights from')
-    parser.add_argument('--gta', '-g', dest='use_gta', action='store_true', help='Generate from GTA testset')
```
`dest` now not needed as argparse names variable based on the string

TheButlah commented on preprocess.py:
```diff
@@ -60,13 +60,17 @@ def process_wav(path):
-                  ('Mu Law', hp.mu_law),
-                  ('Hop Length', hp.hop_length),
-                  ('CPU Count', cpu_count())])
+    n_processes = min(8, cpu_count())
```
Whats the motivation for limiting to 8 cores? My workstations have 40 and 36, wouldn't it make sense to utilize them? If the goal is to avoid using all the theads, why not keep the OS and maybe one other thread free by doing `max(1, cpu_count()-2)`?

fatchord replied:
I'm thinking about situations where there are multiple users of a workstation. I think 8 is plenty - wouldn't there be diminishing returns after that with synchronisation overhead and whatnot?

TheButlah replied:
Hard drive: 
When checking to see how many wav files (13100 files total) are processed in two minutes time, I get the following: 
1/36 cores: 1150
4/36 cores: 2201
8/36 cores: 2932
18/36 cores: 3983
24/36 cores: 4418
35/36 cores: 5067
![cores vs processing time](https://user-images.githubusercontent.com/6969415/61733474-c1a17780-ad34-11e9-946b-687b0d166b6c.png)

I think with that big of a speedup, it makes sense to let the number of processes scale with the number of cores. For multi-user workstations, the OS scheduler should give the threads priority based on the niceness versus other users. If you're really worried about being kind to your coworkers, I think a command line flag should do the trick. 

I do agree that saving one core for the OS to ensure responsiveness is a good call though. I propose we do `n_processes = max(1, args.num_cores if args.num_cores else cpu_count()-1)`

fatchord replied:
Yeah that's good.

TheButlah commented:
In other news, I cannot for the life of me figure out why the flatten parameters warning is still happening. We use `GRUCell` and `LSTMCell`, but those don't extend `RNNBase` and hence don't have the `flatten_parameters()` function. Any idea why PyTorch is complaining @fatchord ? I'm going to move on and fix the errors you mentioned in Tacotron and forget about the parameter flattening for now

TheButlah commented:
OK @fatchord I think I got the bugs sorted out - I've tested the code on both CPU and GPU and Multi-GPU.

There is one issue however. What should the default value of self.r be? Is it 1? Basically this is an issue when training from scratch without a checkpoint file - when this happens, the model just uses a value of 0 for r and then it breaks.

Depending on whether 1 can be used safely as a default or not may affect whether the pretrained models need to be updated. Let me know, for now I will make 1 the default.

fatchord commented:
Yeah, I think we're good to here. Many thanks for your awesome work - I really appreciate it.

TheButlah commented:
I think the branch you merged didn't contain my latest commit. Can you undo the merge and pull in the latest commits?

TheButlah commented:
@fatchord Sorry for force pushing so much as that was the cause of this, I just didn't want to pollute the commit history with a bunch of erroneous commits

TheButlah commented:
OK I've identified some new bugs that break this patch, so unfortunately your master branch is broken right now. I suggest reverting the patch till then.

Bugs that need to be fixed (will keep this list updated):
1. attention tensor is being accessed as a numpy array but its a Tensor. This is because I had to make all returns from `forward()` tensors, but the attention plot used to be just a numpy array.
2. Cosmetic issue in preprocess arguments, fixed on my master but still needs to be merged into fatchord's master.
3. Issue with default r value being 0. Fixed on my master but still needs to be merged into fatchord's master.

TheButlah commented:
Status update: Just did a massive refactoring of the paths system, so that everything uses pathlib. Also began the process of refactoring the model checkpointing code. This is part of the effort to fix the "resuming from training" bug and the potential bugs dealing with model checkpoints, but is also to make the code cleaner, less brittle, and more likely to work on windows. Again as a reminder, @fatchord I suggest unmerging the patch until its ready, as both your repo and mine are in a broken state right now.

fatchord commented:
@TheButlah I've reverted back - let me know what I can do to get this merged.

TheButlah commented:
@fatchord if you can add some code in a new branch on my fork to handle backwards compatability when loading the checkpoints (make sure r and step are loaded), that would be the most helpful :)

fatchord commented:
@TheButlah No problem - I'll get around to it as soon as I get the time.

fatchord commented:
@TheButlah Apologies, I've been super-busy the last week. I'll make time at the weekend to try and get your fork merged.

TheButlah commented:
No problem. As best as I can tell, Tacotron seems to be working properly. I
haven't really worked on wavernn, which is the primary reason I've not
submitted a new PR yet.

On Fri, Aug 2, 2019, 03:08 fatchord <notifications@github.com> wrote:

> @TheButlah <https://github.com/TheButlah> Apologies, I've been super-busy
> the last week. I'll make time at the weekend to try and get your fork
> merged.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fatchord/WaveRNN/pull/109?email_source=notifications&email_token=ABVFQR3FPQIGB2BKVMBQJDLQCQBSHA5CNFSM4ID3ZNGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3NKGNA#issuecomment-517645108>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABVFQR2FFX6ZWYSF2VNBQHTQCQBSHANCNFSM4ID3ZNGA>
> .
>

TheButlah commented:
To clarify, other than backwards compatability with previous saved models, I think I'm pretty happy with the state of the tacotron model on my fork. It properly does multi-gpu, guarantees that resuming checkpoints doesn't break training, and uses pathlib. I haven't tested the generation scripts yet though. 

I'm at the point where I'm in a bit of a crunch time at my job so I probably don't have time to work on WaveRNN right now, but it ought to be fairly easy to finish up the patch so that everything else is working, if either @fatchord or some other contributor can do it

fatchord commented:
@TheButlah That's sounds great. I think the best thing is to train new models since I want to try predicting more mel channels (128 or 256). Once they are trained I will test the things you mention, add new pretrained weights and merge the whole thing. Should be done in a week.

TheButlah commented:
Quick update: My fork now appears to have both WaveRNN, Tacotron, and the respective inference scripts working. Training successfully works on both CPU, GPU, and Multi-GPU.
